core/ledger/kvledger/txmgmt/statedb/statecouchdb/cache.go

```diff
@@ -155,6 +160,14 @@ func (c *cache) getCache(namespace string) *fastcache.Cache {
 	return c.usrCache
 }
 
+func setImplicitValue(c *fastcache.Cache, cacheKey []byte, value []byte) {
+	if cap(value) > fastCacheValueSizeLimit {
+		c.SetBig(cacheKey, value)
+	} else {
+		c.Set(cacheKey, value)
+	}
+}
```

Comment:
what would happen if the oldValue was Big but the new value is not and vice-versa? Will the old big value be deleted before storing the new small value and vice-versa? It isn't clear from the documentation or code. 

```
	c := fastcache.New(64 * 1024 * 1024)

	c.Set([]byte("key1"), []byte("value1"))
	val = c.Get(nil, []byte("key1"))
	fmt.Printf("Get(): Key is set with small value: length [%d], value[%s]\n", len(val), string(val))

	big := make([]byte, 2*64*1024)
	rand.Read(big)
	c.SetBig([]byte("key1"), big)
	val = c.GetBig(nil, []byte("key1"))
	fmt.Printf("GetBig(): Key is set with big value: length [%d]\n", len(val))

	val = c.Get(nil, []byte("key1"))
	fmt.Printf("Get(): Key is set with big value: length [%d], value[%s]\n", len(val), string(val))

	c.Set([]byte("key1"), []byte("value1"))
	val = c.Get(nil, []byte("key1"))
	fmt.Printf("Get(): Key is set with small value: length [%d], value[%s]\n", len(val), string(val))

	val = c.GetBig(nil, []byte("key1"))
	fmt.Printf("GetBig(): Key is set with big value: length [%d]\n", len(val))
```

Here, after setting the big value to an existing key with a small value, the `Get()` does not return `nil` but some random bytes. Though the `GetBig()` is always used in this PR first before the `Get()`, there is a lot of unspecified assumption and hence, the code look fragile. 

There are two options:

1. We can always use `GetBig()` and `SetBig()` irrespective of the size limit but need to benchmark the performance impact for value < 64 KB.
2. What @manish-sethi suggested here -- https://github.com/hyperledger/fabric/pull/1909/files#r492786264 

I would prefer 2. If needed, we can quantify the impact of always using big APIs and make a separate PR if needed for 1.

Reply:
Just to take a note, when looking for alternate ways, I would add one more approach in which we could remember in the cache itself if the key was saved using the `SetBig` api (in a separate namespace) and can evaluate the overheads compared to always using Big apis.

As far as this bug fix is concerned, my inclination also would be to keep it simple as we need to back port this as well. We can maintain the existing behavior (i.e., not storing bigger values) and just fix the bug.

Reply:
I have already upgraded fastcache. Take a look in file changes. Currently we can use `Has(<key>)` method from fastcache and the problem is getting payload from cache(usage BigCache or Fastcache). So I changed this firstly by using `GetBig()` and if length of bytes is less than fastcache capacity then I use `Get()`. This will prevent from getting incorrect byte array.

Feel free for any advice/correction/refactor

Reply:
One thing I didn't mention is that `Has()` works on fastcache and bigcache because Cache is shared between this two

Reply:
This still does not solve the problem that I had highlighted in my previous [comment](https://github.com/hyperledger/fabric/pull/1909#discussion_r492786264). I'll encourage you to run locally the sample test that I posted in my comment to understand the problem with the fix. I am not sure if upgrading fastcache solves the issue. The problem is with the wrong usage of APIs. To elaborate the case, there could be a key that we previously set using the `Set()` api. Now, in your fix, you first look for the key using the `GetBig()` api and as highlighted in my test, this can lead to a panic crash if the value of the key (that is set previously) has certain properties.

For fixing the bug, as we discussed before, in the function `putState`, we can check the size of the value and if greater than the limit, don't call `Set` api and instead call `Del` api in order to remove the previous version of the key from the cache. This maintains the current behavior i.e., we don't store the big values in the cache and just fix the bug.

For allowing for bigger values later, we discussed a couple of options above and they can be tried and evaluated in a separate PR.

Reply:
In the update PR, the usage of `Has()`, `GetBig()`, `Get()`, and comparing the `valueSize` and `sizeLimit` complicates the simple cache usage. Let's go with the following suggestion for a bug fix. 

> For fixing the bug, as we discussed before, in the function putState, we can check the size of the value and if greater than the limit, don't call Set api and instead call Del api in order to remove the previous version of the key from the cache. This maintains the current behavior i.e., we don't store the big values in the cache and just fix the bug.

As storing > 64 KB is a new feature, let's do that separately with less complication.

Reply:
I've already understood that you won't take care about `bigValues` in cache. 

1. You want to save them only in couchDB
2. When `GetState` has been called in statecouchDB try get from cache, if null get it directly from couchDb

Am I correct?

In connection with the new version I thought that using provided method like `Has` and `HasGet` will be nice.
You have seen changes in `UpdateStates` and `getState`method

I did what @manish-sethi  suggested to me.